File: packages/sdk/pact-go/transport/messages.go
package transport

import (
	"bufio"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"strings"
)

type MessageHandler func(context.Context, map[string]any) error

func ParseRPCMessages(rawBody string) ([]map[string]any, error) {
	trimmed := strings.TrimSpace(rawBody)
	if trimmed == "" {
		return []map[string]any{}, nil
	}
	if strings.HasPrefix(trimmed, "{") {
		message, err := parseJSONMessage(trimmed)
		if err != nil {
			return nil, err
		}
		return []map[string]any{message}, nil
	}

	messages := make([]map[string]any, 0)
	buffer := make([]string, 0)
	for _, line := range strings.Split(rawBody, "\n") {
		if strings.TrimSpace(line) == "" {
			if len(buffer) == 0 {
				continue
			}
			message, err := parseJSONMessage(strings.Join(buffer, "\n"))
			if err != nil {
				return nil, err
			}
			messages = append(messages, message)
			buffer = buffer[:0]
			continue
		}
		if strings.HasPrefix(line, "data:") {
			buffer = append(buffer, strings.TrimSpace(line[5:]))
		}
	}
	if len(buffer) > 0 {
		message, err := parseJSONMessage(strings.Join(buffer, "\n"))
		if err != nil {
			return nil, err
		}
		messages = append(messages, message)
	}
	return messages, nil
}

func ReadRPCMessagesUntilTerminal(
	ctx context.Context,
	responseBody io.Reader,
	contentType string,
	expectedID any,
	onMessage MessageHandler,
) ([]map[string]any, error) {
	if onMessage == nil {
		onMessage = func(context.Context, map[string]any) error { return nil }
	}

	if strings.HasPrefix(strings.ToLower(contentType), "text/event-stream") {
		return readStreamMessagesUntilTerminal(ctx, responseBody, expectedID, onMessage)
	}

	rawBody, err := io.ReadAll(responseBody)
	if err != nil {
		return nil, err
	}
	messages, err := ParseRPCMessages(string(rawBody))
	if err != nil {
		return nil, err
	}
	for _, message := range messages {
		if isTerminalMessage(message, expectedID) {
			continue
		}
		if err := onMessage(ctx, message); err != nil {
			return nil, err
		}
	}
	return messages, nil
}

func TerminalMessage(messages []map[string]any, expectedID any) (map[string]any, error) {
	for _, message := range messages {
		if isTerminalMessage(message, expectedID) {
			if rpcErr, ok := message["error"].(map[string]any); ok {
				return nil, fmt.Errorf("json-rpc error: %v", rpcErr["message"])
			}
			return message, nil
		}
	}
	return nil, fmt.Errorf("no terminal response for JSON-RPC id %v", expectedID)
}

func readStreamMessagesUntilTerminal(
	ctx context.Context,
	responseBody io.Reader,
	expectedID any,
	onMessage MessageHandler,
) ([]map[string]any, error) {
	messages := make([]map[string]any, 0)
	buffer := make([]string, 0)
	scanner := bufio.NewScanner(responseBody)
	scanner.Buffer(make([]byte, 0, 64*1024), 1024*1024)

	flush := func() (bool, error) {
		if len(buffer) == 0 {
			return false, nil
		}
		message, err := parseJSONMessage(strings.Join(buffer, "\n"))
		if err != nil {
			return false, err
		}
		messages = append(messages, message)
		buffer = buffer[:0]
		if isTerminalMessage(message, expectedID) {
			return true, nil
		}
		if err := onMessage(ctx, message); err != nil {
			return false, err
		}
		return false, nil
	}

	for scanner.Scan() {
		line := scanner.Text()
		if strings.TrimSpace(line) == "" {
			terminal, err := flush()
			if err != nil {
				return nil, err
			}
			if terminal {
				return messages, nil
			}
			continue
		}
		if strings.HasPrefix(line, "data:") {
			buffer = append(buffer, strings.TrimSpace(line[5:]))
		}
	}
	if err := scanner.Err(); err != nil {
		return nil, err
	}
	if _, err := flush(); err != nil {
		return nil, err
	}
	return messages, nil
}

func parseJSONMessage(input string) (map[string]any, error) {
	decoder := json.NewDecoder(strings.NewReader(input))
	decoder.UseNumber()
	var message map[string]any
	if err := decoder.Decode(&message); err != nil {
		return nil, err
	}
	return message, nil
}

func isTerminalMessage(message map[string]any, expectedID any) bool {
	if expectedID == nil {
		return false
	}
	if _, isNotification := message["method"]; isNotification {
		return false
	}
	return normalizeID(message["id"]) == normalizeID(expectedID)
}

func normalizeID(value any) string {
	switch typed := value.(type) {
	case json.Number:
		return typed.String()
	case string:
		return typed
	case fmt.Stringer:
		return typed.String()
	default:
		return fmt.Sprint(typed)
	}
}

File: sdks/typescript/chio-ts/src/transport/messages.ts
export type JsonRpcId = string | number | null;

export type JsonRpcMessage = Record<string, unknown> & {
  id?: JsonRpcId;
  method?: string;
};

export type RpcMessageHandler = (message: JsonRpcMessage) => void | Promise<void>;

function rpcIdsEqual(left: JsonRpcId | undefined, right: JsonRpcId | undefined): boolean {
  return left === right;
}

function parseJsonRpcMessage(input: string): JsonRpcMessage {
  return JSON.parse(input) as JsonRpcMessage;
}

export function parseRpcMessages(rawBody: string): JsonRpcMessage[] {
  const trimmed = rawBody.trim();
  if (!trimmed) {
    return [];
  }
  if (trimmed.startsWith("{") || trimmed.startsWith("[")) {
    const parsed = JSON.parse(trimmed) as JsonRpcMessage | JsonRpcMessage[];
    return Array.isArray(parsed) ? parsed : [parsed];
  }

  const messages: JsonRpcMessage[] = [];
  let buffer: string[] = [];
  for (const line of rawBody.split(/\r?\n/)) {
    if (!line.trim()) {
      if (buffer.length > 0) {
        messages.push(parseJsonRpcMessage(buffer.join("\n")));
        buffer = [];
      }
      continue;
    }
    if (line.startsWith("data:")) {
      buffer.push(line.slice(5).trimStart());
    }
  }
  if (buffer.length > 0) {
    messages.push(parseJsonRpcMessage(buffer.join("\n")));
  }
  return messages;
}

function isTerminalMessage(message: JsonRpcMessage, expectedId: JsonRpcId | undefined): boolean {
  return expectedId !== undefined && rpcIdsEqual(message.id, expectedId) && !message.method;
}

export async function readRpcMessagesUntilTerminal(
  response: Response,
  expectedId?: JsonRpcId,
  onMessage: RpcMessageHandler = async () => {},
): Promise<JsonRpcMessage[]> {
  if (!response.body) {
    const parsedMessages = parseRpcMessages(await response.text());
    const messages: JsonRpcMessage[] = [];
    for (const message of parsedMessages) {
      messages.push(message);
      if (isTerminalMessage(message, expectedId)) {
        break;
      }
      await onMessage(message);
    }
    return messages;
  }

  const reader = response.body.getReader();
  const decoder = new TextDecoder();
  const messages: JsonRpcMessage[] = [];
  const eventData: string[] = [];
  let rawBody = "";
  let buffer = "";

  while (true) {
    const { value, done } = await reader.read();
    if (done) {
      rawBody += decoder.decode();
      break;
    }

    const chunk = decoder.decode(value, { stream: true });
    rawBody += chunk;
    buffer += chunk;
    const lines = buffer.split(/\r?\n/);
    buffer = lines.pop() ?? "";

    for (const line of lines) {
      if (!line.trim()) {
        if (eventData.length > 0) {
          const message = parseJsonRpcMessage(eventData.join("\n"));
          messages.push(message);
          if (isTerminalMessage(message, expectedId)) {
            await reader.cancel();
            return messages;
          }
          await onMessage(message);
          eventData.length = 0;
        }
        continue;
      }
      if (line.startsWith("data:")) {
        eventData.push(line.slice(5).trimStart());
      }
    }
  }

  if (buffer.trim()) {
    if (buffer.startsWith("data:")) {
      eventData.push(buffer.slice(5).trimStart());
    }
  }

  if (eventData.length > 0) {
    const message = parseJsonRpcMessage(eventData.join("\n"));
    messages.push(message);
    if (!isTerminalMessage(message, expectedId)) {
      await onMessage(message);
    }
  }
  if (messages.length === 0) {
    const parsedMessages = parseRpcMessages(rawBody);
    for (const message of parsedMessages) {
      messages.push(message);
      if (isTerminalMessage(message, expectedId)) {
        return messages;
      }
      await onMessage(message);
    }
  }
  return messages;
}

export function terminalMessage(messages: JsonRpcMessage[], expectedId: JsonRpcId): JsonRpcMessage {
  const match = messages.find((message) => rpcIdsEqual(message.id, expectedId) && !message.method);
  if (!match) {
    throw new Error(`no terminal response for JSON-RPC id ${expectedId}`);
  }
  const error = match.error;
  if (error && typeof error === "object") {
    const message =
      "message" in error && typeof error.message === "string"
        ? error.message
        : `JSON-RPC error for id ${expectedId}`;
    throw new Error(message);
  }
  return match;
}
